src/comgr.cpp

```diff
@@ -989,4 +998,290 @@ void BuildAsm(const std::string& name,
+    }
+
+    void Compile(const std::vector<std::string>& options)
+    {
+        std::vector<const char*> c_options;
+        std::transform(options.begin(),
+                       options.end(),
+                       std::back_inserter(c_options),
+                       [](const std::string& s) { return s.c_str(); });
+        comgr::LogOptions(c_options.data(), c_options.size());
+
+        HIPRTC_CALL_INFO_THROW_MSG(
+            hiprtcCompileProgram(prog.get(), c_options.size(), c_options.data()),
+            src_name,
+            GetLog(true));
+        const auto log = GetLog(false);
+        if(!log.empty())
+            MIOPEN_LOG_I(log);
+    }
+
+    void GetCode(std::vector<char>& bytes) const
+    {
+        std::size_t sz = 0;
+        HIPRTC_CALL_INFO_THROW(hiprtcGetCodeSize(prog.get(), &sz), src_name);
+        bytes.resize(sz);
+        HIPRTC_CALL_INFO_THROW(hiprtcGetCode(prog.get(), &bytes[0]), src_name);
+    }
+
+    private:
+    void LogInputFile(const std::string& name, const std::string& content)
+    {
+        if(miopen::IsEnabled(MIOPEN_DEBUG_COMGR_LOG_SOURCE_NAMES{}))
+            MIOPEN_LOG_I(name << ' ' << content.size() << " bytes");
+        if(miopen::IsLogging(miopen::LoggingLevel::Info))
+        {
+            const auto show_first = miopen::Value(MIOPEN_DEBUG_COMGR_LOG_SOURCE_TEXT{}, 0);
+            if(show_first > 0)
+            {
+                const auto text_length =
+                    (content.size() > show_first) ? show_first : content.size();
+                const std::string text(content, 0, text_length);
+                MIOPEN_LOG_I(text);
+            }
+        }
+    }
+
+    std::string GetLog(const bool error_handling)
+    {
+        std::string text;
+        try
+        {
+            std::size_t n = 0;
+            HIPRTC_CALL_INFO_THROW(hiprtcGetProgramLogSize(prog.get(), &n), n);
+            if(n < 2)
+                return {error_handling ? "warning: HIPRTC error log empty" : ""};
+            std::vector<char> buffer(n);
+            HIPRTC_CALL_INFO_THROW(hiprtcGetProgramLog(prog.get(), buffer.data()), n);
+            assert(buffer.back() == 0);
+            return {buffer.begin(), buffer.end() - 1};
+        }
+        catch(Error&)
+        {
+            if(error_handling)
+                return {"HIPRTC error: failed to get error log"};
+            throw;
+            /// \ref catch_and_rethrow_in_getlog
+        }
+    }
+};
+
+void BuildHip(const std::string& name,
+              const std::string& text,
+              const std::string& options,
+              const miopen::TargetProperties& target,
+              std::vector<char>& binary)
+{
+    PrintVersion();
+    try
+    {
+        auto opts =
+            miopen::SplitSpaceSeparated(options, miopen::comgr::compiler::lc::GetOptionsNoSplit());
+        compiler::lc::RemoveOptionsUnwanted(opts);
+        opts.push_back("-DWORKAROUND_ISSUE_HIPRTC_TRUE_TYPE"); // Workaround for SWDEV-308073
```

Comment:
@JehandadKhan 
> so that we something needs to be fixed, it only needs to be fixed in one place.

I do not understand. Can you please clarify the use case?